apps/src/lib/node/ledger/mod.rs

```diff
@@ -1,425 +1,214 @@
+            }
+            Request::EndBlock(end) => match BlockHeight::try_from(end.height) {
+                Ok(height) => Ok(Response::EndBlock(self.end_block(height))),
+                Err(_) => {
+                    tracing::error!("Unexpected block height {}", end.height);
+                    Ok(Response::EndBlock(Default::default()))
                 }
-
-                let gas = gas::as_i64(result.gas_used);
-                (gas, Ok(result))
+            },
+            Request::Commit(_) => Ok(Response::Commit(self.commit())),
+            Request::Flush(_) => Ok(Response::Flush(Default::default())),
+            Request::SetOption(_) => {
+                Ok(Response::SetOption(Default::default()))
             }
-            err @ Err(_) => {
-                let gas =
-                    gas::as_i64(self.gas_meter.get_current_transaction_gas());
-                (gas, err)
+            Request::Echo(msg) => Ok(Response::Echo(response::Echo {
+                message: msg.message,
+            })),
+            Request::CheckTx(tx) => {
+                let r#type = match CheckTxType::from_i32(tx.r#type)
+                    .expect("received unexpected CheckTxType from ABCI")
+                {
+                    CheckTxType::New => MempoolTxType::NewTransaction,
+                    CheckTxType::Recheck => MempoolTxType::RecheckTransaction,
+                };
+                Ok(Response::CheckTx(self.mempool_validate(&*tx.tx, r#type)))
             }
-        }
+            Request::ListSnapshots(_) => {
+                Ok(Response::ListSnapshots(Default::default()))
+            }
+            Request::OfferSnapshot(_) => {
+                Ok(Response::OfferSnapshot(Default::default()))
+            }
+            Request::LoadSnapshotChunk(_) => {
+                Ok(Response::LoadSnapshotChunk(Default::default()))
+            }
+            Request::ApplySnapshotChunk(_) => {
+                Ok(Response::ApplySnapshotChunk(Default::default()))
+            }
+        };
+        tracing::info!(?rsp);
+        Box::pin(async move { rsp.map_err(|e| e.into()) }.boxed())
     }
+}
 
-    /// Simulate validation and application of a transaction.
-    fn dry_run_tx(&mut self, tx_bytes: &[u8]) -> Result<String> {
-        let mut gas_meter = BlockGasMeter::default();
-        let mut write_log = self.write_log.clone();
-        let result = protocol::apply_tx(
-            tx_bytes,
-            &mut gas_meter,
-            &mut write_log,
-            &self.storage,
+pub fn reset(config: config::Ledger) -> Result<(), shell::Error> {
+    shell::reset(config)
+}
+
+#[tokio::main]
+async fn run_shell(
+    config: config::Ledger,
+    abort_registration: AbortRegistration,
+) {
+    // Construct our ABCI application.
+    let service = Shell::new(&config.db, config::DEFAULT_CHAIN_ID.to_owned());
+
+    // Split it into components.
+    let (consensus, mempool, snapshot, info) = split::service(service, 1);
```

Comment:
the `bound` arg can probably be greater to allow serving more concurrent requests, iiuc

Reply:
Ok, trying to pick a sensible value here. According to `tower`:
```
When Buffer’s implementation of poll_ready returns Poll::Ready, it reserves a slot in the channel 
for the forthcoming call. However, if this call doesn’t arrive, this reserved slot may be held up for a 
long time. As a result, it’s advisable to set bound to be at least the maximum number of concurrent 
requests the Buffer will see. If you do not, all the slots in the buffer may be held up by futures that
have just called poll_ready but will not issue a call, which prevents other senders from issuing new
requests.
```
I think our max number if 4 in this case, so I'm going to tentatively suggest 5. Apparently wildly overshooting it can hide backpressure according to the tower-abci doc string
```
However, large buffers hide backpressure from propagating to the caller.
```